### PR TITLE
[SPARK-18433][SQL] Improve DataSource option keys to be more case-insensitive

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

This PR aims to improve DataSource option keys to be more case-insensitive

DataSource partially use CaseInsensitiveMap in code-path. For example, the following fails to find url.

```scala
val df = spark.createDataFrame(sparkContext.parallelize(arr2x2), schema2)
df.write.format("jdbc")
    .option("UrL", url1)
    .option("dbtable", "TEST.SAVETEST")
    .options(properties.asScala)
    .save()
```

This PR makes DataSource options to use CaseInsensitiveMap internally and also makes DataSource to use CaseInsensitiveMap generally except `InMemoryFileIndex` and `InsertIntoHadoopFsRelationCommand`. We can not pass them CaseInsensitiveMap because they creates new case-sensitive HadoopConfs by calling newHadoopConfWithOptions(options) inside.

## How was this patch tested?

Pass the Jenkins test with newly added test cases.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala [hunk 1]
```diff
@@ -34,35 +34,42 @@ private[sql] class JSONOptions(
     @transient private val parameters: Map[String, String])
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala]: can we just force users to pass a `CaseInsensitiveMap`?

Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala]: Oh, sure! You mean `require(parameters.isInstanceOf[CaseInsensitiveMap])`, right?

Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala]: I will update this PR in that way.

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala]: `JSONOptions` is spark private, we can just declare the type is `CaseInsensitiveMap`, i.e. `parameters: CaseInsensitiveMap`

**Comment by dongjoon-hyun** [comment]
Thank you, @cloud-fan .
It becomes much better. Could you review again please?

**Comment by dongjoon-hyun** [comment]
Rebased to resolve conflicts.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala [hunk 2]
```diff
@@ -1366,7 +1366,7 @@ class JsonSuite extends QueryTest with SharedSQLContext with TestJsonData {
 
   test("SPARK-6245 JsonRDD.inferSchema on empty RDD") {
     // This is really a test that it doesn't throw an exception
-    val emptySchema = InferSchema.infer(empty, "", new JSONOptions(Map()))
+    val emptySchema = InferSchema.infer(empty, "", new JSONOptions(Map.empty[String, String]))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: unnecessary change?

**Comment by cloud-fan** [comment]
LGTM

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala (hunk 2 above)
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/json/JsonSuite.scala]: In this case, we expect `Map[String,String]`, but `Map()` or `Map.empty` is `Map[Nothing,Nothing]`.
So, there was some compilation issue.

**Comment by cloud-fan** [comment]
thanks, merging to master/2.1!

**Comment by dongjoon-hyun** [comment]
Thank you so much, @cloud-fan ! Also, thank you for this issue, @gatorsmile .
